tests/semantic_analyser.cpp

```diff
@@ -8,237 +11,150 @@
+  template <Options options = Options{}, typename... Ts>
+    requires((std::is_same_v<std::decay_t<Ts>, Mock> ||
+              std::is_same_v<std::decay_t<Ts>, Warning> ||
+              std::is_same_v<std::decay_t<Ts>, NoWarning> ||
+              std::is_same_v<std::decay_t<Ts>, Error> ||
+              std::is_same_v<std::decay_t<Ts>, ExpectedAST>) &&
+             ...)
+  ast::ASTContext test(std::string_view input, Ts &&...args)
+  {
+    ast::ASTContext ast("stdin", std::string(clean_prefix(input)));
 
-ast::ASTContext test(const std::string &input)
-{
-  auto bpftrace = get_mock_bpftrace();
-  return test(*bpftrace, true, input, 0, {}, true, false);
-}
+    // Reset for each iteration. We only guarantee that the types remain
+    // valid after the ASTContext has been returned.
+    bpftrace_ = get_mock_bpftrace();
 
-ast::ASTContext test(BPFtrace &bpftrace,
-                     const std::string &input,
-                     std::string_view expected_ast)
-{
-  auto ast = test(bpftrace, true, input, 0, {}, true, false);
+    // Extract all extra arguments.
+    auto mock = extract<Mock>(args...);
+    auto warning = extract<Warning>(args...);
+    auto nowarning = extract<NoWarning>(args...);
+    auto error = extract<Error>(args...);
+    auto expected_ast = extract<ExpectedAST>(args...);
 
-  if (expected_ast[0] == '\n')
-    expected_ast.remove_prefix(1); // Remove initial '\n'
+    if (!mock) {
+      mock.emplace(*bpftrace_);
+    }
+    mock->bpftrace.safe_mode_ = options.safe_mode;
+    mock->bpftrace.feature_ = std::make_unique<MockBPFfeature>(
+        options.features);
+    if constexpr (options.with_child) {
+      mock->bpftrace.cmd_ = "not-empty"; // Used by SemanticAnalyser.
+    }
 
-  std::ostringstream out;
-  ast::Printer printer(out);
-  printer.visit(ast.root);
+    auto ok = ast::PassManager()
+                  .put(ast)
+                  .put(mock->bpftrace)
+                  .add(CreateParsePass())
+                  .add(ast::CreateMacroExpansionPass())
+                  .add(ast::CreateParseAttachpointsPass())
+                  .add(ast::CreateFieldAnalyserPass())
+                  .add(ast::CreateClangParsePass())
+                  .add(ast::CreateCMacroExpansionPass())
+                  .add(ast::CreateFoldLiteralsPass())
+                  .add(ast::CreateMapSugarPass())
+                  .add(ast::CreateSemanticPass())
+                  .run();
+    EXPECT_TRUE(bool(ok));
+
+    std::stringstream out;
+    ast.diagnostics().emit(out, ast::Diagnostics::Severity::Warning);
+    if (warning && !warning->str.empty()) {
+      EXPECT_THAT(out.str(), HasSubstr(clean_prefix(warning->str)))
+          << out.str();
+    }
+    if (nowarning && !nowarning->str.empty()) {
+      EXPECT_THAT(out.str(), Not(HasSubstr(clean_prefix(nowarning->str))))
+          << out.str();
+    }
+    out.str("");
+    ast.diagnostics().emit(out, ast::Diagnostics::Severity::Error);
+    if (error && !error->str.empty()) {
```

Comment:
Actually I do see `Error{}` used below but then if `error->str` is empty shouldn't we assert that `out.str()` is not empty?

Reply:
Ah, good point.